allennlp/training/cv_trainer.py

```diff
@@ -0,0 +1,204 @@
+import copy
+import logging
+import os
+from typing import Any, Dict, Iterable, List, Optional, Tuple, Type
+
+import numpy as np
+from overrides import overrides
+from sklearn.model_selection import BaseCrossValidator, GroupKFold, KFold, LeaveOneGroupOut, LeaveOneOut, \
+    LeavePGroupsOut, LeavePOut, StratifiedKFold, TimeSeriesSplit
+import torch
+
+from allennlp.common import Params, Registrable
+from allennlp.common.util import dump_metrics
+from allennlp.data import DataIterator, Instance
+from allennlp.models.model import Model
+from allennlp.training import util as training_util
+from allennlp.training.metrics import Average
+from allennlp.training.no_op_trainer import NoOpTrainer
+from allennlp.training.trainer import Trainer, TrainerPieces
+from allennlp.training.trainer_base import TrainerBase
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+class CrossValidationSplitter(Registrable):
+    def __init__(self, cross_validator: BaseCrossValidator) -> None:
+        super().__init__()
+        self.cross_validator = cross_validator
+
+    def __call__(self, instances: List[Instance], groups: Optional[Any] = None) -> Iterable[Tuple[np.ndarray,
+                                                                                                  np.ndarray]]:
+        return self.cross_validator.split(instances, groups=groups)
+
+    def get_n_splits(self, instances: List[Instance], groups: Optional[Any] = None) -> int:
+        return self.cross_validator.get_n_splits(instances, groups=groups)
+
+
+class _CrossValidationSplitterWrapper:
+    def __init__(self, cross_validator_class: Type[BaseCrossValidator]) -> None:
+        super().__init__()
+        self.cross_validator_class = cross_validator_class
+
+    def from_params(self, params: Params) -> CrossValidationSplitter:
+        cross_validator = self.cross_validator_class(**params.as_dict())
```

Comment:
[According to the Python docs](https://docs.python.org/3/library/typing.html#typing.Type), `cross_validator_score` should have as type all the classes that have different constructors. In this case, the base constructor doesn't accept args while the subclasses do. So, strictly speaking, type checking will fail with some tools. However, I preferred to keep it simple (the change would be to accept multiple types, probably to list every subclass).